src/System.Devices.Gpio/System.Devices.Gpio/RaspberryPiDriver.cs

```diff
@@ -0,0 +1,828 @@
+            //[FieldOffset(0x88)]
+            //public UInt32 GPAFEN0;
+
+            /////<summary>GPIO Pin Async. Falling Edge Detect 1, 32 bits, R/W</summary>
+            //[FieldOffset(0x8C)]
+            //public UInt32 GPAFEN1;
+
+            //[FieldOffset(0x90)]
+            //- Reserved - - 
+
+            ///<summary>GPIO Pin Pull-up/down Enable, 32 bits, R/W</summary>
+            [FieldOffset(0x94)]
+            public UInt32 GPPUD;
+
+            ///<summary>GPIO Pin Pull-up/down Enable Clock, 2x32 bits, R/W</summary>
+            [FieldOffset(0x98)]
+            public fixed UInt32 GPPUDCLK[2];
+
+            /////<summary>GPIO Pin Pull-up/down Enable Clock 0, 32 bits, R/W</summary>
+            //[FieldOffset(0x98)]
+            //public UInt32 GPPUDCLK0;
+
+            /////<summary>GPIO Pin Pull-up/down Enable Clock 1, 32 bits, R/W</summary>
+            //[FieldOffset(0x9C)]
+            //public UInt32 GPPUDCLK1;
+
+            //[FieldOffset(0xA0)]
+            //- Reserved - - 
+
+            //[FieldOffset(0xB0)]
+            //- Test 4 bits R/W
+        }
+
+        #endregion
+
+        #region Interop
+
+        private const string LibraryName = "libc";
+
+        [Flags]
+        private enum FileOpenFlags
+        {
+            O_RDWR = 0x02,
+            O_SYNC = 0x101000
+        }
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int open([MarshalAs(UnmanagedType.LPStr)] string pathname, FileOpenFlags flags);
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int close(int fd);
+
+        [Flags]
+        private enum MemoryMappedProtections
+        {
+            PROT_NONE = 0x0,
+            PROT_READ = 0x1,
+            PROT_WRITE = 0x2,
+            PROT_EXEC = 0x4
+        }
+
+        [Flags]
+        private enum MemoryMappedFlags
+        {
+            MAP_SHARED = 0x01,
+            MAP_PRIVATE = 0x02,
+            MAP_FIXED = 0x10
+        }
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern IntPtr mmap(IntPtr addr, int length, MemoryMappedProtections prot, MemoryMappedFlags flags, int fd, int offset);
+
+        [DllImport(LibraryName, SetLastError = true)]
+        private static extern int munmap(IntPtr addr, int length);
+
+        #endregion
+
+        private const string GpioMemoryFilePath = "/dev/gpiomem";
+        private const int GpioBaseOffset = 0;
+
+        private RegisterView* _registerViewPointer = null;
+
+        public RaspberryPiDriver()
+        {
+            var fileDescriptor = open(GpioMemoryFilePath, FileOpenFlags.O_RDWR | FileOpenFlags.O_SYNC);
+
+            if (fileDescriptor < 0)
+            {
+                throw new IOException($"open error number: {Marshal.GetLastWin32Error()}");
+            }
+
+            //Console.WriteLine($"file descriptor = {fileDescriptor}");
```

Comment:
The info from this type of debug helpers should be available via strace. Any problem to run it on the device? An alternative is to use the trace mechanisms from CoreFx or debug output if they are valuable only for debug runs.

Reply:
I cannot debug in the device.